common/tasks/interleaved_weighted_round_robin.go

```diff
@@ -0,0 +1,255 @@
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package tasks
+
+import (
+	"sort"
+	"sync"
+	"sync/atomic"
+
+	"go.temporal.io/server/common"
+	"go.temporal.io/server/common/log"
+	"go.temporal.io/server/common/metrics"
+)
+
+type (
+	// InterleavedWeightedRoundRobinSchedulerOptions is the config for
+	// interleaved weighted round robin scheduler
+	InterleavedWeightedRoundRobinSchedulerOptions struct {
+		QueueSize   int
+		WorkerCount int
+	}
+
+	InterleavedWeightedRoundRobinScheduler struct {
+		status int32
+		option InterleavedWeightedRoundRobinSchedulerOptions
+
+		processor    Processor
+		metricsScope metrics.Scope
+		logger       log.Logger
+
+		notifyChan   chan struct{}
+		shutdownChan chan struct{}
+
+		sync.RWMutex
+		priorityToWeight     map[int]int
+		weightToTaskChannels map[int]*WeightedChannel
+		iwrrChannels         []*WeightedChannel
+	}
+)
+
+func NewInterleavedWeightedRoundRobinScheduler(
+	option InterleavedWeightedRoundRobinSchedulerOptions,
+	priorityToWeight map[int]int,
+	processor Processor,
+	metricsClient metrics.Client,
+	logger log.Logger,
+) *InterleavedWeightedRoundRobinScheduler {
+	return &InterleavedWeightedRoundRobinScheduler{
+		status: common.DaemonStatusInitialized,
+		option: option,
+
+		processor:    processor,
+		metricsScope: metricsClient.Scope(metrics.TaskSchedulerScope),
+		logger:       logger,
+
+		notifyChan:   make(chan struct{}, 1),
+		shutdownChan: make(chan struct{}),
+
+		priorityToWeight:     priorityToWeight,
+		weightToTaskChannels: make(map[int]*WeightedChannel),
+		iwrrChannels:         []*WeightedChannel{},
+	}
+}
+
+func (s *InterleavedWeightedRoundRobinScheduler) Start() {
+	if !atomic.CompareAndSwapInt32(
+		&s.status,
+		common.DaemonStatusInitialized,
+		common.DaemonStatusStarted,
+	) {
+		return
+	}
+
+	s.processor.Start()
+
+	go s.eventLoop()
+
+	s.logger.Info("interleaved weighted round robin task scheduler started")
+}
+
+func (s *InterleavedWeightedRoundRobinScheduler) Stop() {
+	if !atomic.CompareAndSwapInt32(
+		&s.status,
+		common.DaemonStatusStarted,
+		common.DaemonStatusStopped,
+	) {
+		return
+	}
+
+	close(s.shutdownChan)
+
+	s.processor.Stop()
+
+	s.drainTasks()
+
+	s.logger.Info("interleaved weighted round robin task scheduler stopped")
+}
+
+func (s *InterleavedWeightedRoundRobinScheduler) Submit(
+	task PriorityTask,
+
+) {
+	channel := s.getOrCreateTaskChannel(s.priorityToWeight[task.GetPriority()])
+	channel.Chan() <- task
+	s.notifyDispatcher()
+}
+
+func (s *InterleavedWeightedRoundRobinScheduler) eventLoop() {
+	for {
+		select {
+		case <-s.notifyChan:
+			for s.hasRemainingTasks() {
+				weightedChannels := s.channels()
+				s.dispatchTasks(weightedChannels)
+			}
```

Comment:
this whole thing should be handled within s.dispatchTasks() which should not take any input. We probably don't need hasRemainingTasks() and s.channels().

Reply:
the reason to split the logic into few pieces
* easier to test each functionality
* do not hold the internal lock for too long

Reply:
This logic belong to dispatchTasks(), you can certainly do it all in dispatchTasks() without having to hold the lock all the time. You could also break logics in to smaller methods. But it would be clean and nice to call s.dispatchTasks() from eventLoop().

Reply:
>This logic belong to dispatchTasks(), you can certainly do it all in dispatchTasks() without having to hold the lock all the time. 

overall effect are the same, plus harder to test